### ISSUE TITLE
Create TOML documents within Terraform

### ISSUE THREAD

**Issue by MeNsaaH** [hashicorp/terraform#28328]
### Current Terraform Version

```
...
```

### Use-cases

I was working on deploying some charts using terraform, and one of the charts values require a toml template. It would be nice to have this in terraform, ability to convert from terraform maps/arrays to toml structure.

**Comment by apparentlymart** [comment]
Hi @MeNsaaH! Thanks for this feature request.

For each of the other formats that we've implemented encoders/decoders for in Terraform, we've had to first complete a design effort to understand which parts of the data model of the particular language correspond with each of the Terraform language types, and also note any situations where there isn't a suitable correspondence and devise a way to deal with that.

For example, when we designed `yamlencode` we had to devise what became the mapping table of types in [the `yamldecode` documentation](https://www.terraform.io/docs/language/functions/yamlencode.html) and also make some subjective decisions about situations where YAML gives multiple options, such as when to use flow vs. block style, how to lay out indented constructs, etc.

I'm not particularly familiar with TOML so I don't know yet how involved such an effort would be, but I do see [the TOML v1.0.0 spec](https://toml.io/en/v1.0.0) which does look like it has all of the information needed to begin a design effort like that.

I have a few observations just from an initial scan, although I've not thought hard about what the tradeoffs for these might be yet:
* TOML has two number types that are both limited in range: "Integer" means 64-bit signed integer (−2^63 to 2^63−1) and "Float" means a 64-bit float. This means that there are valid `number` values in Terraform that would not be serializable in TOML, so we'd need to define what happens in that case.
* Terraform has no types corresponding with TOML's Offset Date-Time, Local Date-Time, Local Date, and Local Time types. This means that unless we can define some special mechanism it wouldn't be possible to generate a TOML-based format that included values of those types.
* TOML has various string serializations, so I imagine a TOML encoder for Terraform would need to use some heuristics to decide for a particular string which exact syntax to use in order to minimize the amount of escaping noise while still keeping the exact value intact.
* Unlike `jsonencode` and `yamlencode`, `tomlencode` would need to require that its direct argument is always an object-typed value, because TOML only allows a table at the root.
* Choosing between "Table" and "Inline Table" seems like another situation where we'd need to rely on some heuristics because Terraform can't know from the input alone what the intent is and thus which serialization would best. It seems like "Inline Table" is only appropriate for relatively-small objects that would therefore not generate a very long line.
* It seems like there are some constraints on what combinations are valid when nesting arrays inside tables, tables inside arrays, tables inside tables, and arrays inside arrays, but I've not analyzed all of the rules in detail to see if that's true. If there are combinations that are not possible then presumably `tomlencode` would need to reject them with an error.

Designing and implementing something like this can be deceptively complex and time-consuming (the `yamlencode` function took several design iterations and a few implementation iterations too, for example), so we've typically been cautious about adding new ones. JSON and YAML ended up being justified due to how much external software consumes values in those formats, but this is the first time we've seen a request for TOML, and so we don't have examples to draw from here.

With that said, it'd be helpful if you could say a little more about what software you are interacting with that expects TOML. Not only will that help us to understand the potential impact of implementing and maintaining a function like this (to trade against the cost of doing it), but having real examples to look at will likely help with making design decisions like the ones I described above that seem like they will require choosing some heuristics.

I expect we won't be able to act on this in the very near future because our focus is in other areas right now. However, if you or anyone else would like to share example use-cases (ideally including real examples of the sort of TOML document you'd like to generate) then that would help with evaluating this feature request and, if we move forward with it, with the design process.

Thanks!

**Comment by MeNsaaH** [comment]
Okay, my use case here is using Gitlab Runner's helm chart. It was recently updated to pass config via toml config instead of the regular yaml see https://gitlab.com/gitlab-org/charts/gitlab-runner/-/blob/master/values.yaml#L167.

I do under the complexity of implementing this. At the moment, I'm using string templates to overcome this hurdle. However, I feel any native approach will be better. I totally agree with you on having to go from JSON to TOML instead of directly moving to TOML. I believe that'll be easier and help avoid having to run through heuristics.

**Comment by apparentlymart** [comment]
Thanks for sharing the example! YAML with TOML embedded inside it is an extra level of :grimacing: that I wasn't expecting, but I guess that's how things go sometimes. :sweat_smile:

**Comment by MaciekLeks** [comment]
It would be nice to have tomencode/tomldecode while TOML becomes more and more popular. I do not know why but it is. I also struggle with kubernetes helm gitlab runner and even have been trying to bring some logic into it, so that's why I've [forked](https://github.com/MaciekLeks/terraform-kubernetes-gitlab-runner) existing project, but I get lost when it comes to affinity aspects inside TOML part of that magician mixture of YAML and TOML.

**Comment by highb** [comment]
> 

Oh no, not necessarily that someone would use TF to interop with Rust crate metadata, just that as a configuration language in general, it is gaining popularity due to it's use there. That leads me to _speculate_ that there will be increasing demand for TOML support from Rust developers over time, as new applications and deployment tools etc. start to get written and adopted that would likely use TOML.

**Comment by crw** [comment]
Thank you for your continued interest in this issue. 

[Terraform version 1.8](https://github.com/hashicorp/terraform/blob/v1.8.0-beta1/CHANGELOG.md) launches with support of provider-defined functions. It is now possible to implement your own functions! We would love to see this implemented as a provider-defined function. 

Please see the [provider-defined functions documentation](https://developer.hashicorp.com/terraform/plugin/framework/functions) to learn how to implement functions in your providers. If you are new to provider development, learn how to [create a new provider](https://developer.hashicorp.com/terraform/plugin/framework) with the Terraform Plugin Framework. If you have any questions, please visit the [Terraform Plugin Development category in our official forum](https://discuss.hashicorp.com/c/terraform-providers/tf-plugin-sdk/43). 

We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!

**Comment by crw** [comment]
(Re-reading this thread, I'm not entirely sure this will be a solution, but keeping the comment as an informational post.)

**Comment by highb** [comment]
> We hope this feature unblocks future function development and provides more flexibility for the Terraform community. Thank you for your continued support of Terraform!

Provider-defined functions should allow someone from the community to contribute `toml_encode` and `toml_decode` functions. As I don't work with Terraform anymore, that someone isn't me but I wish someone else luck with the implementation!

**Comment by apparentlymart** [comment]
There is at least one provider whose documentation suggests it would meet this use-case: [`tobotimus/toml`](https://registry.terraform.io/providers/tobotimus/toml/latest/docs).

I've not tried this provider myself and so I cannot vouch for what it does, but the goal of provider-defined functions was to allow for providers like this to reduce the pressure to add every single serialization format to Terraform Core, and it seems like that worked out here. Thanks to @tobotimus for publishing this provider.

Since we don't intend to add TOML-specific functions as builtin and there is now at least one provider that seems to address the use-case, I'm going to close this issue. Thanks!
